### DISCUSSION TITLE
Multiple device support ideas

### DISCUSSION THREAD

**Discussion by joeahn95** [espressif/esp-iot-solution#446] [Q&A]
Currently utilizing the `usb_device_uac` component as a speaker / microphone device. I'm wondering if there is a straightforward way to also classify my device as an HID, CDC, MSC, etc. I see that espressif offers the esp_tinyusb component for support on these devices (excluding audio), so I'm wondering if anyone has experience on utilizing these components for multiple device support. Are there limitations that require me to utilize the tinyusb component directly? 

Thanks for your help.

**Comment by lijunru-hub** [comment]
For composite devices, it is recommended to create your own device based on TinyUSB. You can refer to this example, which packages all the files for a UAC device into a single folder. This allows you to easily integrate UAC functionality into your own composite device.

https://github.com/espressif/esp-iot-solution/tree/master/examples/usb/device/usb_extend_screen/main/uac

We also have a document to assist you in developing your own device based on TinyUSB.

https://docs.espressif.com/projects/esp-iot-solution/en/latest/usb/usb_overview/tinyusb_development_guide.html

**Reply by joeahn95** [reply]
Thank you! A few follow up questions:

- I'm noticing that the usb_extend_screen overall project depends on `leeebo/tinyusb`, rather than the master tinyusb source component. I noticed that `usb_device_uac` also depends on `leeebo/tinyusb` as well. Do the UAC functions specifically rely on this specific leebo branch of tinyusb, or can we switch to the master tinyusb source? Asking because examples for other device functionalities (such as HID) seem to depend straight on the master tinyusb source if the idf version is >= 5.0. (see https://github.com/espressif/esp-iot-solution/blob/master/examples/usb/device/usb_hid_device/main/idf_component.yml) 

- The project you referred to seems to be designed for esp32p4. I'm guessing that the audio component from this project should still work for esp32s3, am I correct?

Thanks again.

**Reply by joeahn95** [reply]
An update, I was able to create a composite device with UAC + HID. My main goal for the HID component is to detect when the host pc is starting or ending a call, such as a zoom or skype meeting. Is there a way to do that via HID? So far, I can only send HID reports that type key strokes or move the mouse.

**Reply by lijunru-hub** [reply]
Applicable to S3, corresponding TinyUSB version uses leeebo/tinyusb, mainly for compiling on IDF v4.4.

HID cannot access the start and end of a meeting unless you create a custom host driver and use a custom HID report to the device side.

**Reply by joeahn95** [reply]
Great thanks, I made some progress regarding a custom HID report.

With an HID keyboard device, I'm able to detect output reports on using the sample descriptor. For example, if I set CAPS LOCK active manually from my pc (and not my HID keyboard), then my device will detect that a `tud_hid_set_report_cb` has been called to show that data has been sent from the host to the device (since the caps lock data was set by the host pc). 

I was able to setup a HID telephony device type and set one output report description. In this case, I want to detect if the host pc starts and stops a meeting. I see in the HID documentation that there is actually a telephony device usage ID for this, called `host call active (0xF2)`. However, when I start a zoom meeting or google meeting, I do not detect any `tud_hid_set_report_cb` to show that the host is currently in a call.

I have my custom HID descriptor here. Do you know if there is anything that looks wrong here? If nothing looks wrong, am I missing any steps? Maybe the `host call active` usage ID is not applicable to online meetings?

```
#define TUD_HID_REPORT_DESC_TELEPHONY(...) \
  HID_USAGE_PAGE ( HID_USAGE_PAGE_TELEPHONY )                     ,\
  HID_USAGE      ( 5 )                  ,\
  HID_COLLECTION ( HID_COLLECTION_APPLICATION )                   ,\
    /* Report ID if any */\
    __VA_ARGS__ \
    /* Host Call Active */ \
    HID_USAGE_PAGE ( HID_USAGE_PAGE_TELEPHONY )                    ,\
      /* 242 is the Host Call Active Usage ID */
      HID_USAGE      ( 242 )      ,\ 
      HID_LOGICAL_MIN( 0                                      )    ,\
      HID_LOGICAL_MAX( 1                                      )    ,\
      HID_REPORT_COUNT( 1                                      )    ,\
      HID_REPORT_SIZE ( 1                                      )    ,\
      HID_OUTPUT      ( HID_DATA | HID_VARIABLE | HID_ABSOLUTE )    ,\
      /* 7 bit padding */ \
      HID_REPORT_COUNT( 1                                      )    ,\
      HID_REPORT_SIZE ( 7                                      )    ,\
      HID_OUTPUT      ( HID_CONSTANT                           )    ,\
  HID_COLLECTION_END \
```
